Framework/PythonInterface/plugins/algorithms/IntegratePeaksShoeboxTOF.py

```diff
@@ -0,0 +1,687 @@
+        center_slice = tuple(slice(nedge, -nedge) for nedge in (np.asarray(kernel.shape) - 1) // 2)
+        edge_mask[center_slice] = False
+        intens_over_sig[edge_mask] = 0
+    return intens_over_sig
+
+
+def find_nearest_peak_in_data_window(data, ispecs, x, ws, peaks, ipk, irow, icol, ix, threshold=2):
+    ipks_near, ispecs_peaks = find_ipks_in_window(ws, peaks, ispecs, ipk, tof_min=x.min(), tof_max=x.max())
+    if len(ipks_near) > 0:
+        tofs = peaks.column("TOF")
+        # get position of nearby peaks in data array in fractional coords
+        shape = np.array(data.shape)
+        pos_near = [
+            np.r_[np.argwhere(ispecs == ispecs_peaks[ii])[0], np.argmin(abs(x - tofs[ipk_near]))] / shape
+            for ii, ipk_near in enumerate(ipks_near)
+        ]
+        # sort data in descending order and select strongest data point nearest to pk (in fractional coordinates)
+        isort = list(zip(*np.unravel_index(np.argsort(-data, axis=None), data.shape)))
+        pk_pos = np.array([irow, icol, ix]) / np.array(data.shape)
+        imax_nearest = None
+        for ibin in isort:
+            if data[ibin] > threshold:
+                # calc distance to predicted peak position
+                bin_pos = np.array(ibin) / shape
+                pk_dist_sq = np.sum((pk_pos - bin_pos) ** 2)
+                for pos in pos_near:
+                    if np.sum((pos - bin_pos) ** 2) > pk_dist_sq:
+                        imax_nearest = ibin
+                        break
+                else:
+                    continue  # executed if inner loop did not break (i.e. pixel closer to nearby peak than this peak)
+                break  # execute if inner loop did break and else branch ignored (i.e. found bin closest to this peak)
+        return imax_nearest  # could be None if no peak found
+    else:
+        # no nearby peaks - return position of maximum in data
+        return np.unravel_index(np.argmax(data), data.shape)
+
+
+def find_ipks_in_window(ws, peaks, ispecs, ipk, tof_min=None, tof_max=None):
+    ispecs_peaks = ws.getIndicesFromDetectorIDs([int(p.getDetectorID()) for p in peaks])
+    ipks_near = np.isin(ispecs_peaks, ispecs)
+    if tof_min and tof_max:
+        tofs = peaks.column("TOF")
+        ipks_near = np.logical_and.reduce((tofs >= tof_min, tofs <= tof_max, ipks_near))
+    ipks_near = np.flatnonzero(ipks_near)  # convert to index
+    ipks_near = np.delete(ipks_near, np.where(ipks_near == ipk))  # remove the peak of interest
+    return ipks_near, ispecs_peaks
+
+
+def integrate_shoebox_at_pos(y, esq, kernel, ipos, weak_peak_threshold, det_edges=None):
+    slices = tuple(
+        [
+            slice(
+                np.clip(ii - kernel.shape[idim] // 2, a_min=0, a_max=y.shape[idim]),
+                np.clip(ii + kernel.shape[idim] // 2 + 1, a_min=0, a_max=y.shape[idim]),
+            )
+            for idim, ii in enumerate(ipos)
+        ]
+    )
```

Comment:
In a couple of places in the class you use this list comprehension to initialize the slice. I agree list comprehension is more elegant way of doing it, but since you do it few times, it may improve on readability if this is part of a separate function on the class with a comment of what it does?

Reply:
Yep, I can try and put it in the function, but if you look at the detail of the slices made they are all different - so I think it would not be possible to generalise without adding a few if/else which may make it less easy to read! Is that OK?

Reply:
You are right. Then is fine, we can keep this as it is. Is not really an issue.